core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -930,30 +941,46 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
       auditContext.setSrcInode(inode);
 
       List<FileInfo> ret = new ArrayList<>();
-      if (inode.isDirectory()) {
-        TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath);
-        try {
-          mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, inodePath);
-        } catch (AccessControlException e) {
-          auditContext.setAllowed(false);
-          throw e;
-        }
-        for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
-          child.lockReadAndCheckParent(inode);
+      listStatusRecursive(ret, inodePath, auditContext, listStatusOptions.isRecursive());
+
+      auditContext.setSucceeded(true);
+      Metrics.FILE_INFOS_GOT.inc();
+      return ret;
+    }
+  }
+
+  private void listStatusRecursive(List<FileInfo> statusList, LockedInodePath originalInodePath,
+                                   AuditContext auditContext, boolean recursive)
+      throws UnavailableException, FileDoesNotExistException,
+      AccessControlException, InvalidPathException {
+    Queue<LockedInodePath> queue = new LinkedList<>();
+    queue.add(originalInodePath);
+
+    try (InodeLockList lockList = new InodeLockList()) {
+      while (!queue.isEmpty()) {
+        LockedInodePath currInodePath = queue.remove();
+        Inode<?> inode = currInodePath.getInode();
+        if (inode.isDirectory()) {
           try {
-            // the path to child for getPath should already be locked.
-            tempInodePath.setDescendant(child, mInodeTree.getPath(child));
-            ret.add(getFileInfoInternal(tempInodePath));
-          } finally {
-            child.unlockRead();
+            mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, originalInodePath);
+          } catch (Exception e) {
+            auditContext.setAllowed(false);
+            throw e;
+          }
+          if (recursive || currInodePath.equals(originalInodePath)) {
+            for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
+              lockList.lockReadAndCheckParent(child, inode);
+              TempInodePathForDescendant tempInodePath
+                  = new TempInodePathForDescendant(originalInodePath);
+              tempInodePath.setDescendant(child, mInodeTree.getPath(child));
+              queue.add(tempInodePath);
```

Comment:
What do you think if we depth-first (lock child, recursive call, unlock) instead of breadth-first? I am asking because we may be able to unlock descendants earlier, which can allow more concurrency?

I think the output might be more meaningful too?

Reply:
done